agent/engine/docker_task_engine.go

```diff
@@ -1534,25 +1535,91 @@ func (engine *DockerTaskEngine) pullAndUpdateContainerReference(task *apitask.Ta
 		defer clearCreds()
 	}
 
-	metadata := engine.client.PullImage(engine.ctx, container.Image, container.RegistryAuthentication, engine.cfg.ImagePullTimeout)
+	imageRef := container.Image
+	imageDigest := container.GetImageDigest()
+	if imageDigest != "" {
+		// If image digest is available then prepare a canonical reference to be used for image pull.
+		// This ensures that the image version referenced by the digest is pulled.
+		canonicalRef, err := referenceutil.GetCanonicalRef(imageRef, imageDigest)
+		if err != nil {
+			logger.Error("Failed to prepare a canonical reference. Cannot pull image.", logger.Fields{
+				field.TaskID:      task.GetID(),
+				field.Container:   container.Name,
+				field.Image:       imageRef,
+				field.ImageDigest: imageDigest,
+				field.Error:       err,
+			})
+			return dockerapi.DockerContainerMetadata{
+				Error: dockerapi.CannotPullContainerError{
+					FromError: fmt.Errorf(
+						"failed to prepare a canonical reference with image '%s' and digest '%s': %w",
+						imageRef, imageDigest, err),
+				},
+			}
+		}
+		imageRef = canonicalRef.String()
+		logger.Info("Prepared a canonical reference for image pull", logger.Fields{
+			field.TaskID:      task.GetID(),
+			field.Container:   container.Name,
+			field.Image:       container.Image,
+			field.ImageDigest: imageDigest,
+			field.ImageRef:    imageRef,
+		})
+	}
+
+	metadata := engine.client.PullImage(engine.ctx, imageRef, container.RegistryAuthentication, engine.cfg.ImagePullTimeout)
 
 	// Don't add internal images(created by ecs-agent) into imagemanger state
 	if container.IsInternal() {
 		return metadata
 	}
 	pullSucceeded := metadata.Error == nil
+
+	if pullSucceeded && imageRef != container.Image {
```

Comment:
Q: Will the `imageRef != container.Image` check ever be false? Looking at [pullImage](https://github.com/amogh09/amazon-ecs-agent/blob/a5f8d25f66fbf4f8839d5e7e5fcbc635a4d75c5e/agent/dockerclient/dockerapi/docker_client.go#L374), it doesn't seem like we're updating either of the two? (please correct me if this is wrong!)

Reply:
It will be false if `container.Image` is already in canonical form like `busybox@sha256:05a79c7279f71f86a2a0d05eb72fcb56ea36139150f0a75cd87e80a4272e4e39` or if  image digest was not resolved for the container (if Docker engine version is too old for digest resolution, for example) and `container.GetImageDigest()` returns an empty string.

Reply:
Sync'd offline! Essentially we're making sure that the image ref that we're using to pull is in canonical reference format will be retagged so that it can be referred to by just the image name or <image_name>:latest. Thanks Amogh!